File: lingting-http/src/main/java/live/lingting/framework/http/HttpClient.java
package live.lingting.framework.http;

import live.lingting.framework.stream.BytesInputStream;
import live.lingting.framework.stream.FileCloneInputStream;
import live.lingting.framework.util.FileUtils;
import live.lingting.framework.util.StreamUtils;
import live.lingting.framework.util.ThreadUtils;
import live.lingting.framework.value.LazyValue;
import lombok.SneakyThrows;

import javax.net.SocketFactory;
import javax.net.ssl.HostnameVerifier;
import javax.net.ssl.SSLContext;
import javax.net.ssl.X509TrustManager;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.CookieManager;
import java.net.CookiePolicy;
import java.net.CookieStore;
import java.net.ProxySelector;
import java.net.URI;
import java.security.KeyManagementException;
import java.security.NoSuchAlgorithmException;
import java.time.Duration;
import java.util.Set;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.atomic.AtomicLong;
import java.util.function.Consumer;

/**
 * @author lingting 2024-09-02 15:28
 */
public abstract class HttpClient {

	/**
	 * @see jdk.internal.net.http.common.Utils#getDisallowedHeaders()
	 */
	protected static final Set<String> HEADERS_DISABLED = Set.of("connection", "content-length", "expect", "host",
			"upgrade");

	public static final File TEMP_DIR = FileUtils.createTempDir("http");

	/**
	 * 默认大小以内文件直接放内存里面
	 */
	public static final long MAX_BYTES = 5242880;

	protected CookieStore cookie;

	public static JavaHttpClient.Builder java() {
		return new JavaHttpClient.Builder();
	}

	public static OkHttpClient.Builder okhttp() {
		return new OkHttpClient.Builder();
	}

	public static InputStream wrap(InputStream source) throws IOException {
		return wrap(source, MAX_BYTES);
	}

	public static InputStream wrap(InputStream source, long maxBytes) throws IOException {
		if (source == null) {
			return new ByteArrayInputStream(new byte[0]);
		}
		AtomicLong size = new AtomicLong(0);

		ByteArrayOutputStream byteOut = new ByteArrayOutputStream();
		LazyValue<FileOutputStream> fileOutValue = new LazyValue<>(() -> null);
		LazyValue<File> fileValue = new LazyValue<>(() -> {
			File file = FileUtils.createTemp(".wrap", TEMP_DIR);
			fileOutValue.set(new FileOutputStream(file));
			return file;
		});
		StreamUtils.read(source, (bytes, len) -> {
			if (!fileValue.isFirst() || size.addAndGet(len) > maxBytes) {
				if (fileValue.isFirst()) {
					fileValue.get();
					fileOutValue.get().write(byteOut.toByteArray());
				}
				fileOutValue.get().write(bytes, 0, len);
			}
			else {
				byteOut.write(bytes, 0, len);
			}
		});
		if (fileValue.isFirst()) {
			return new BytesInputStream(byteOut.toByteArray());
		}
		return new FileCloneInputStream(fileValue.get());
	}

	public CookieStore cookie() {
		return cookie;
	}

	public abstract Object client();

	public abstract HttpResponse request(HttpRequest request) throws IOException;

	public abstract void request(HttpRequest request, ResponseCallback callback) throws IOException;

	public <T> T request(HttpRequest request, Class<T> cls) throws IOException {
		HttpResponse response = request(request);
		return response.convert(cls);
	}

	public HttpResponse get(URI uri) throws IOException {
		return request(HttpRequest.builder().get().url(uri).build());
	}

	public abstract static class Builder<C extends HttpClient, B extends Builder<C, B>> {

		protected ExecutorService executor = ThreadUtils.executor();

		protected boolean redirects = true;

		protected SocketFactory socketFactory;

		/**
		 * HostnameVerifier，用于HTTPS安全连接
		 */
		protected HostnameVerifier hostnameVerifier;

		/**
		 * 用于HTTPS安全连接
		 */
		protected SSLContext sslContext;

		protected X509TrustManager trustManager;

		protected Duration callTimeout;

		protected Duration connectTimeout;

		protected Duration readTimeout;

		protected Duration writeTimeout;

		protected ProxySelector proxySelector;

		protected CookieStore cookie;

		public B socketFactory(SocketFactory socketFactory) {
			this.socketFactory = socketFactory;
			return (B) this;
		}

		public B executor(ExecutorService executor) {
			this.executor = executor;
			return (B) this;
		}

		public B redirects(boolean redirects) {
			this.redirects = redirects;
			return (B) this;
		}

		public B hostnameVerifier(HostnameVerifier hostnameVerifier) {
			this.hostnameVerifier = hostnameVerifier;
			return (B) this;
		}

		public B ssl(X509TrustManager trustManager) throws NoSuchAlgorithmException, KeyManagementException {
			SSLContext context = Https.sslContext(trustManager);
			return ssl(context, trustManager);
		}

		public B ssl(SSLContext context, X509TrustManager trustManager) {
			this.sslContext = context;
			this.trustManager = trustManager;
			return (B) this;
		}

		@SneakyThrows
		public B disableSsl() {
			X509TrustManager manager = Https.SSL_DISABLED_TRUST_MANAGER;
			HostnameVerifier verifier = Https.SSL_DISABLED_HOSTNAME_VERIFIER;
			return ssl(manager).hostnameVerifier(verifier);
		}

		public B callTimeout(Duration callTimeout) {
			this.callTimeout = callTimeout;
			return (B) this;
		}

		public B connectTimeout(Duration connectTimeout) {
			this.connectTimeout = connectTimeout;
			return (B) this;
		}

		public B readTimeout(Duration readTimeout) {
			this.readTimeout = readTimeout;
			return (B) this;
		}

		public B writeTimeout(Duration writeTimeout) {
			this.writeTimeout = writeTimeout;
			return (B) this;
		}

		/**
		 * 无限等待时间
		 */
		public B infiniteTimeout() {
			return timeout(Duration.ZERO, Duration.ZERO, Duration.ZERO, Duration.ZERO);
		}

		public B timeout(Duration connectTimeout, Duration readTimeout) {
			return connectTimeout(connectTimeout).readTimeout(readTimeout);
		}

		public B timeout(Duration callTimeout, Duration connectTimeout, Duration readTimeout, Duration writeTimeout) {
			return callTimeout(callTimeout).connectTimeout(connectTimeout)
				.readTimeout(readTimeout)
				.writeTimeout(writeTimeout);
		}

		public B proxySelector(ProxySelector proxySelector) {
			this.proxySelector = proxySelector;
			return (B) this;
		}

		public B cookie(CookieStore cookie) {
			this.cookie = cookie;
			return (B) this;
		}

		public B memoryCookie() {
			CookieManager manager = new CookieManager(null, CookiePolicy.ACCEPT_ALL);
			return cookie(manager.getCookieStore());
		}

		protected <A> void nonNull(A a, Consumer<A> consumer) {
			if (a == null) {
				return;
			}
			consumer.accept(a);
		}

		public C build() {
			C c = doBuild();
			c.cookie = cookie;
			return c;
		}

		protected abstract C doBuild();

	}

}

File: lingting-http/src/main/kotlin/live/lingting/framework/http/HttpClient.kt
package live.lingting.framework.http

import live.lingting.framework.util.FileUtils
import live.lingting.framework.util.ThreadUtils
import java.io.File
import java.net.CookieManager
import java.net.CookiePolicy
import java.net.CookieStore
import java.net.ProxySelector
import java.net.URI
import java.time.Duration
import java.util.concurrent.ExecutorService
import java.util.function.Consumer
import javax.net.SocketFactory
import javax.net.ssl.HostnameVerifier
import javax.net.ssl.SSLContext
import javax.net.ssl.X509TrustManager

/**
 * @author lingting 2024-09-02 15:28
 */
@Suppress("UNCHECKED_CAST")
abstract class HttpClient {

    companion object {
        /**
         * @see jdk.internal.net.http.common.Utils.getDisallowedHeaders
         */
        @JvmField
        val HEADERS_DISABLED: Set<String> = setOf(
            "connection", "content-length", "expect", "host",
            "upgrade"
        )

        @JvmField
        val TEMP_DIR: File = FileUtils.createTempDir("http")

        @JvmStatic
        fun builder() = default()

        fun default(): Builder<*, *> = okhttp()

        @JvmStatic
        fun java(): JavaHttpClient.Builder {
            return JavaHttpClient.Builder()
        }

        @JvmStatic
        fun okhttp(): OkHttpClient.Builder {
            return OkHttpClient.Builder()
        }

        fun createFile() = FileUtils.createTemp(".wrap", TEMP_DIR)

    }

    var memoryResponse = true

    protected var cookie: CookieStore? = null

    fun cookie(): CookieStore? {
        return cookie
    }

    abstract fun client(): Any

    abstract fun request(request: HttpRequest): HttpResponse

    abstract fun request(request: HttpRequest, callback: ResponseCallback)

    fun <T> request(request: HttpRequest, cls: Class<T>): T {
        val response = request(request)
        return response.convert(cls)
    }

    fun get(uri: URI): HttpResponse {
        return request(HttpRequest.builder().get().url(uri).build())
    }

    fun get(url: String): HttpResponse {
        return request(HttpRequest.builder().get().url(url).build())
    }

    fun <T> get(url: String, cls: Class<T>): T? {
        val builder = HttpRequest.builder().get().url(url)
        return request(builder.build(), cls)
    }

    abstract class Builder<C : HttpClient, B : Builder<C, B>> {

        protected var executor: ExecutorService? = ThreadUtils

        protected var redirects: Boolean = true

        protected var socketFactory: SocketFactory? = null

        /**
         * HostnameVerifier，用于HTTPS安全连接
         */
        protected var hostnameVerifier: HostnameVerifier? = null

        /**
         * 用于HTTPS安全连接
         */
        protected var sslContext: SSLContext? = null

        protected var trustManager: X509TrustManager? = null

        /**
         * java 实现无效
         */
        protected var callTimeout: Duration? = null

        protected var connectTimeout: Duration? = null

        protected var readTimeout: Duration? = null

        protected var writeTimeout: Duration? = null

        protected var proxySelector: ProxySelector? = null

        protected var cookie: CookieStore? = null

        protected var memoryResponse: Boolean = true

        fun socketFactory(socketFactory: SocketFactory?): B {
            this.socketFactory = socketFactory
            return this as B
        }

        fun executor(executor: ExecutorService): B {
            this.executor = executor
            return this as B
        }

        fun redirects(redirects: Boolean): B {
            this.redirects = redirects
            return this as B
        }

        fun hostnameVerifier(hostnameVerifier: HostnameVerifier?): B {
            this.hostnameVerifier = hostnameVerifier
            return this as B
        }

        fun ssl(trustManager: X509TrustManager): B {
            val context = Https.sslContext(trustManager)
            return ssl(context, trustManager)
        }

        fun ssl(context: SSLContext?, trustManager: X509TrustManager?): B {
            this.sslContext = context
            this.trustManager = trustManager
            return this as B
        }

        fun disableSsl(): B {
            val manager = Https.SSL_DISABLED_TRUST_MANAGER
            val verifier = Https.SSL_DISABLED_HOSTNAME_VERIFIER
            return ssl(manager).hostnameVerifier(verifier)
        }

        /**
         * java 实现无效
         */
        fun callTimeout(callTimeout: Duration?): B {
            this.callTimeout = callTimeout
            return this as B
        }

        fun connectTimeout(connectTimeout: Duration?): B {
            this.connectTimeout = connectTimeout
            return this as B
        }

        fun readTimeout(readTimeout: Duration?): B {
            this.readTimeout = readTimeout
            return this as B
        }

        fun writeTimeout(writeTimeout: Duration?): B {
            this.writeTimeout = writeTimeout
            return this as B
        }

        /**
         * 无限等待时间
         */
        open fun infiniteTimeout(): B {
            return timeout(Duration.ZERO, Duration.ZERO, Duration.ZERO, Duration.ZERO)
        }

        fun timeout(connectTimeout: Duration?, readTimeout: Duration?): B {
            return connectTimeout(connectTimeout).readTimeout(readTimeout)
        }

        fun timeout(
            callTimeout: Duration?,
            connectTimeout: Duration?,
            readTimeout: Duration?,
            writeTimeout: Duration?
        ): B {
            return callTimeout(callTimeout).connectTimeout(connectTimeout)
                .readTimeout(readTimeout)
                .writeTimeout(writeTimeout)
        }

        fun proxySelector(proxySelector: ProxySelector?): B {
            this.proxySelector = proxySelector
            return this as B
        }

        fun cookie(cookie: CookieStore?): B {
            this.cookie = cookie
            return this as B
        }

        fun memoryCookie(): B {
            val manager = CookieManager(null, CookiePolicy.ACCEPT_ALL)
            return cookie(manager.cookieStore)
        }

        fun memoryResponse(): B {
            memoryResponse = true
            return this as B
        }

        fun fileResponse(): B {
            memoryResponse = false
            return this as B
        }

        protected fun <A> nonNull(a: A?, consumer: Consumer<A>) {
            if (a == null) {
                return
            }
            consumer.accept(a)
        }

        fun build(): C {
            val c = doBuild()
            c.cookie = cookie
            return c
        }

        protected abstract fun doBuild(): C
    }

}
